File: sliding-window/121_best_time_to_buy_and_sell_stock.py
"""
Problem: Best Time to Buy and Sell Stock
LeetCode #: 121
Difficulty: Easy
Link: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

Approach: Track minimum price seen so far and calculate max profit.
Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float("inf")
        max_profit = 0

        for price in prices:
            min_price = min(min_price, price)
            max_profit = max(max_profit, price - min_price)

        return max_profit


# --- Test ---
if __name__ == "__main__":
    sol = Solution()
    print(sol.maxProfit([7, 1, 5, 3, 6, 4]))  # 5
    print(sol.maxProfit([7, 6, 4, 3, 1]))      # 0

File: sliding-window/121_best_time_to_buy_and_sell_stock.java
/*
 * Problem: Best Time to Buy and Sell Stock
 * LeetCode #: 121
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
 *
 * Approach: Track minimum buy price seen so far, update max profit at each step.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

public class Solution {
    public int maxProfit(int[] prices) {
        int minPrice = Integer.MAX_VALUE;
        int maxProfit = 0;
        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price;
            } else if (price - minPrice > maxProfit) {
                maxProfit = price - minPrice;
            }
        }
        return maxProfit;
    }

    // --- Test ---
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.maxProfit(new int[]{7, 1, 5, 3, 6, 4})); // 5
    }
}
